File: TopCoder/src/main/java/com/leetcode/String/RegularExpressionMatching.java
package com.leetcode.String;

public class RegularExpressionMatching {

    public boolean isMatch(String s, String p) {

        if (p.length() == 0) {
            return s.length() == 0;
        }
        if (s.length() == 0 && p.length() == 0) {
            return true;
        }
        if(p.length() > 1 && p.charAt(1) == '*') {
            return isMatch(s, p.substring(2)) || s.length() > 0 && ((s.charAt(0) == p.charAt(0) || '.' == p.charAt(0))
                                                  && isMatch(s.substring(1), p));
        }

        if (s.length() > 0  && ((s.charAt(0) == p.charAt(0)) || p.charAt(0) == '.')) {
            return isMatch(s.substring(1), p.substring(1));
        }
        return false;
    }


    public boolean dp(String s, String p) {
        boolean[][] dp = new boolean[s.length()+1][p.length()+1];
        dp[0][0] = true;
        int rowLength = s.length() + 1;
        int colLength = p.length() + 1;

        for (int i = 1 ; i < rowLength ; i++) {
            dp[i][0] = false;
        }

        for (int j= 1 ; j < colLength ; j++) {
            if (j > 1 && p.charAt(j-1) == '*') {
                dp[0][j] = dp[0][j-2];
            }
        }

        for (int i = 1 ; i < rowLength ; i++) {
            for (int j = 1 ; j < colLength ; j++) {
                if (p.charAt(j-1) == '*') {
                    if ( j > 1 ) {
                        dp[i][j] = dp[i][j - 2];
                    }
                    if ( j > 1 && (p.charAt(j-2) == s.charAt(i-1) || p.charAt(j-2) == '.')) {
                        dp[i][j] = dp[i][j] || dp[i-1][j];
                    }
                } else if (s.charAt(i-1) == p.charAt(j-1) || p.charAt(j-1) == '.') {
                    dp[i][j] = dp[i-1][j-1];
                }
            }
        }

        return dp[rowLength-1][colLength-1];
    }
    public static void main(String[] args) {
        String s = "aaa";
        String p = "ab*a*c*a";
        boolean res = new RegularExpressionMatching().dp(s, p);
        System.out.println(res);
    }

}

File: TopCoder/python/misc/RegularExpressionMatching.py
class Solution:
    def isMatch(self, s, p):
        rowlength = len(s) + 1
        collength = len(p) + 1
        table = [[False]*collength for _ in range(rowlength)]
        table[0][0] = True

        for i in range(2, collength):
            table[0][i] = table[0][i-2] and p[i-1] == '*'

        for i in range(1, rowlength):
            for j in range(1, collength):
                if p[j-1] == '*':
                    if j > 1:
                        table[i][j] = table[i][j-2]

                    if p[j-2] == '.' or s[i-1] == p[j-2]:
                        table[i][j] = table[i][j] or table[i-1][j-1]
                elif p[j-1] == s[i-1] or p[j-1] == '.':
                    table[i][j] = table[i-1][j-1]

        return table[rowlength-1][collength-1]

sol = Solution()
res = sol.isMatch('ab', '.*')
print(res)